libbeat/processors/syslog/syslog.go

```diff
@@ -0,0 +1,178 @@
+
+// config defines the configuration for this processor.
+type config struct {
+	Field         string            `config:"field" validate:"required"`
+	Format        syslog.Format     `config:"format"`
+	TimeZone      *cfgtype.Timezone `config:"timezone"`
+	OverwriteKeys bool              `config:"overwrite_keys"`
+	IgnoreMissing bool              `config:"ignore_missing"`
+	IgnoreFailure bool              `config:"ignore_failure"`
+}
+
+// processor defines a syslog processor.
+type processor struct {
+	config
+
+	log   *logp.Logger
+	stats processorStats
+}
+
+// processorStats contains the metrics fields for the syslog processor.
+type processorStats struct {
+	// Success measures the number of successfully parsed syslog messages.
+	Success *monitoring.Int
+	// Failure measures the number of occurrences where a message was unable to be parsed.
+	Failure *monitoring.Int
+	// Missing measures the number of occurrences where an event was missing the required input field.
+	Missing *monitoring.Int
+}
+
+// init will register various aspects of this processor.
+func init() {
+	processors.RegisterPlugin(procName,
+		checks.ConfigChecked(New,
+			checks.RequireFields(
+				"field",
+			),
+			checks.AllowedFields(
+				"field",
+				"format",
+				"timezone",
+				"overwrite_keys",
+				"ignore_missing",
+				"ignore_failure",
+			),
+		),
+	)
+	jsprocessor.RegisterPlugin("Syslog", New)
+}
+
+// defaultConfig will return a config with default values.
+func defaultConfig() config {
+	return config{
+		Field:         "message",
+		Format:        syslog.FormatAuto,
+		TimeZone:      cfgtype.MustNewTimezone("Local"),
+		OverwriteKeys: true,
+	}
+}
+
+// New creates a new processor from the provided configuration, or an error if the configuration is invalid.
+func New(c *common.Config) (processors.Processor, error) {
+	cfg := defaultConfig()
+
+	if err := c.Unpack(&cfg); err != nil {
+		return nil, fmt.Errorf("fail to unpack the "+procName+" processor configuration: %w", err)
+	}
+
+	id := int(instanceID.Inc())
+	log := logp.NewLogger(logName).With("instance_id", id)
+	registry := monitoring.Default.NewRegistry(logName+"."+strconv.Itoa(id), monitoring.DoNotReport)
+
+	return &processor{
+		config: cfg,
+		log:    log,
+		stats: processorStats{
+			Success: monitoring.NewInt(registry, "success"),
+			Failure: monitoring.NewInt(registry, "failure"),
+			Missing: monitoring.NewInt(registry, "missing"),
+		},
+	}, nil
+}
+
+// Run will process an event and will update the fields based on the parsed message, or an error if the
+// message could not be parsed. If an error occurs and the configuration is set to not ignore errors,
+// the 'error.message' field will be set with error that was encountered.
+func (p *processor) Run(event *beat.Event) (*beat.Event, error) {
+	if err := p.run(event); err != nil && !p.IgnoreFailure {
+		err = fmt.Errorf(procName+" failed to process field %q: %w", p.Field, err)
+		_, _ = event.PutValue("error.message", err.Error())
```

Comment:
This is where the failure semantics of Beats processors are broken IMO. The interface definition is weak so all of the processors do different things w.r.t. error handling, and we have this behavior where errors returned by processors **do not** stop the processing. It's too late (like by 3yrs) to change the behavior, but we will fix it for Agent v2.

The library implementation that does stop on error, but some refactoring happened and this code isn't used in the real pipeline.
https://github.com/elastic/beats/blob/a3fe79648592169bf81d482f5d3895a5045d140a/libbeat/processors/processor.go#L192-L195

This is the implementation that's actually used:
https://github.com/elastic/beats/blob/405c3428b9f5c135c78058aa2c183feb3200edd9/libbeat/publisher/processing/processors.go#L127-L135

If you were to use `syslog` within a `script` processor then the error would stop the pipeline and [automatically write](https://github.com/elastic/beats/blob/7930f9ed0f75696bfc8feff4400a896232184231/libbeat/processors/script/javascript/session.go#L243-L244) to `error.message` (or make `error.message` an array if it already existed).
https://github.com/elastic/beats/blob/a3fe79648592169bf81d482f5d3895a5045d140a/libbeat/processors/script/javascript/module/processor/chain.go#L190-L192

So what to do here? My advice is to append if it already exists just so that this processor is not responsible for overwriting errors. Ideally the caller would be responsible for handling the error and adding it to `error.message`, but that's not the case here.